### PR TITLE
cache mountpoints in the db

### PR DESCRIPTION
**Pull request by icewind1991**
~~~(Nothing is tested atm besides the registering of mounts)~~~

cc @nickvergessen @PVince81 @DeepDiver1975

### CODE REVIEW

**Comment by PVince81** [comment]
Are we really **caching** in the DB or is it a permanent store ? I don't want to make the same mistake like with oc_filecache where people believe it's a cache when it's not :smile:

File: db_structure.xml [hunk 1]
```diff
@@ -129,6 +129,91 @@
...
+			</field>
+
+			<field>
+				<name>storage_id</name>
+				<type>integer</type>
+				<notnull>true</notnull>
+			</field>
+
+			<field>
+				<name>root_id</name>
```
Comment (PVince81) [inline db_structure.xml]: is that the file id ?

Maybe add a comment with "// foreign key: oc_filecache.fileid"

File: db_structure.xml [hunk 2]
```diff
@@ -129,6 +129,91 @@
 
 	<table>
 
+		<name>*dbprefix*mounts</name>
```
Comment (PVince81) [inline db_structure.xml]: Later on please add a comment about what this table is about

**Comment by PVince81** [comment]
@Xenopathic check this out, you might like it too :smile:  (WIP)

**Comment by PVince81** [comment]
Thanks a lot @icewind1991

**Comment by PVince81** [comment]
Ref https://github.com/owncloud/core/issues/15166

**Comment by icewind1991** [comment]
> Are we really caching in the DB or is it a permanent store ? I don't want to make the same mistake like with oc_filecache where people believe it's a cache when it's not 

I think 'cache' is the correct name since it's calculated from whatever the mount providers give us

**Comment by icewind1991** [comment]
Now actually working and with tests

**Comment by icewind1991** [comment]
@nickvergessen @PVince81 please review

File: lib/private/files/config/mountprovidercollection.php [hunk 3]
```diff
@@ -77,4 +86,21 @@ public function registerProvider(IMountProvider $provider) {
...
 		$this->emit('\OC\Files\Config', 'registerMountProvider', [$provider]);
 	}
+
+	/**
+	 * Cache mounts for user
+	 *
+	 * @param IUser $user
+	 * @param IMountPoint[] $mountPoints
+	 */
+	public function cacheMounts(IUser $user, array $mountPoints) {
```
Comment (PVince81) [inline lib/private/files/config/mountprovidercollection.php]: why not just also `registerMounts` ? (I don't mind, just asking)

File: lib/private/files/config/usermountcache.php [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+use OCP\Files\Config\ICachedMountInfo;
+use OCP\Files\Config\IUserMountCache;
+use OCP\Files\Mount\IMountPoint;
+use OCP\ICache;
+use OCP\IDBConnection;
+use OCP\ILogger;
+use OCP\IUser;
+use OCP\IUserManager;
+
+class UserMountCache implements IUserMountCache {
```
Comment (PVince81) [inline lib/private/files/config/usermountcache.php]: PHPDoc ?

File: lib/private/files/config/usermountcache.php [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+	 * @param ILogger $logger
+	 */
+	public function __construct(IDBConnection $connection, IUserManager $userManager, ILogger $logger) {
+		$this->connection = $connection;
+		$this->userManager = $userManager;
+		$this->logger = $logger;
+	}
+
+	public function registerMounts(IUser $user, array $mounts) {
+		$mounts = array_filter($mounts, function (IMountPoint $mount) {
```
Comment (PVince81) [inline lib/private/files/config/usermountcache.php]: When you do data gymnastics, it might help to add a few comments to explain what you are doing, to save time for readers.

Comment (PVince81) [inline lib/private/files/config/usermountcache.php]: So does this get all mounts that have a cache and excludes the ones that don't ?

File: lib/private/files/config/usermountcache.php [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+			return $mount->getStorage()->getCache();
+		});
+		$mounts = array_values($mounts);
+		/** @var ICachedMountInfo[] $newMounts */
+		$newMounts = array_map(function (IMountPoint $mount) use ($user) {
+			$storage = $mount->getStorage();
+			$rootId = (int)$storage->getCache()->getId('');
+			$storageId = (int)$storage->getStorageCache()->getNumericId();
+			if ($rootId === -1) {
```
Comment (PVince81) [inline lib/private/files/config/usermountcache.php]: what? why?

Comment (PVince81) [inline lib/private/files/config/usermountcache.php]: (add comment to clarify)

File: lib/private/files/config/usermountcache.php [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+		/** @var ICachedMountInfo[] $removedMounts */
+		$removedMounts = array_udiff($cachedMounts, $newMounts, $mountDiff);
+
+		foreach ($addedMounts as $mount) {
+			$this->addToCache($mount);
+			$this->mountsForUsers[$user->getUID()][] = $mount;
+		}
+		foreach ($removedMounts as $mount) {
+			$this->removeFromCache($mount);
+			$this->mountsForUsers[$user->getUID()] = [];
```
Comment (PVince81) [inline lib/private/files/config/usermountcache.php]: Why are you resetting the user's mount to an empty array in the loop ? This instruction doesn't seem to be related to the loop.

Would it be better to use "unset()" instead to avoid confusing "user has no mounts" with "user mounts haven't been retrieved yet" ? (if applicable)

File: lib/private/files/config/usermountcache.php [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+		}
+	}
+
+	private function addToCache(ICachedMountInfo $mount) {
+		$builder = $this->connection->getQueryBuilder();
+
+		$query = $builder->insert('mounts')
+			->values([
+				'storage_id' => ':storage',
```
Comment (PVince81) [inline lib/private/files/config/usermountcache.php]: @nickvergessen once told me to use "$query->createNamedParameter()", not sure which usage is better though.

File: lib/private/files/config/usermountcache.php [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+			':storage' => $mount->getStorageId(),
+			':root' => $mount->getRootId(),
+			':user' => $mount->getUser()->getUID(),
+			':mount' => $mount->getMountPoint()
+		]);
+		try {
+			$query->execute();
+		} catch (UniqueConstraintViolationException $e) {
+			// seems to mainly happen in tests
+			$this->logger->error('Duplicate entry while inserting mount');
```
Comment (PVince81) [inline lib/private/files/config/usermountcache.php]: Do we need some kind of locking or transaction to avoid concurrency issues ?

File: lib/private/files/filesystem.php [hunk 10]
```diff
@@ -413,7 +415,8 @@ public static function initMountPoints($user = '') {
 			$homeStorage['arguments']['legacy'] = true;
 		}
 
-		self::mount($homeStorage['class'], $homeStorage['arguments'], $user);
```
Comment (PVince81) [inline lib/private/files/filesystem.php]: Do we need to modify "self::mount" to also use the same logic as you used before ? Or remove this method if not used anywhere else ?

File: lib/public/files/config/icachedmountinfo.php [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+namespace OCP\Files\Config;
+
+use OCP\Files\Node;
+use OCP\IUser;
+
+/**
+ * @since 9.0.0
+ */
+interface ICachedMountInfo {
```
Comment (PVince81) [inline lib/public/files/config/icachedmountinfo.php]: PHPDoc, what is this ?

File: lib/public/files/config/imountprovidercollection.php [hunk 12]
```diff
@@ -22,6 +22,7 @@
 
 namespace OCP\Files\Config;
 
+use OCP\Files\Mount\IMountPoint;
```
Comment (PVince81) [inline lib/public/files/config/imountprovidercollection.php]: Suspicious use without actual usage ?

File: lib/public/files/config/iusermountcache.php [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+namespace OCP\Files\Config;
+
+use OCP\Files\Mount\IMountPoint;
+use OCP\IUser;
+
+/**
+ * @since 9.0.0
+ */
+interface IUserMountCache {
```
Comment (PVince81) [inline lib/public/files/config/iusermountcache.php]: PHPDoc ?

File: tests/lib/files/config/usermountcache.php [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+use OC\Files\Storage\Temporary;
+use OC\Log;
+use OC\User\Manager;
+use OCP\Files\Config\ICachedMountInfo;
+use OCP\IDBConnection;
+use OCP\IUserManager;
+use Test\TestCase;
+use Test\Util\User\Dummy;
+
+class NullLogger extends Log {
```
Comment (PVince81) [inline tests/lib/files/config/usermountcache.php]: Please put this into a separate file. Maybe make it a general test utility class.

Only one class per file.

File: tests/lib/files/config/usermountcache.php [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+	private $cache;
+
+	public function setUp() {
+		$this->connection = \OC::$server->getDatabaseConnection();
+		$this->userManager = new Manager(null);
+		$userBackend = new Dummy();
+		$userBackend->createUser('u1', '');
+		$userBackend->createUser('u2', '');
+		$this->userManager->registerBackend($userBackend);
+		$this->cache = new \OC\Files\Config\UserMountCache($this->connection, $this->userManager, new NullLogger());
```
Comment (PVince81) [inline tests/lib/files/config/usermountcache.php]: Why not use a mock instead of NullLogger ?

File: tests/lib/files/config/usermountcache.php [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+		$this->assertEquals($user, $cachedMount->getUser());
+		$this->assertEquals($storage->getCache()->getId(''), $cachedMount->getRootId());
+		$this->assertEquals($storage->getStorageCache()->getNumericId(), $cachedMount->getStorageId());
+	}
+
+	public function testSameMounts() {
+		$user = $this->userManager->get('u1');
+
+		$storage = $this->getStorage(10, 20);
+		$mount = new MountPoint($storage, '/asd/');
```
Comment (PVince81) [inline tests/lib/files/config/usermountcache.php]: Does creating a mount without trailing slash auto-adds the trailing slash ? We had cases in the past where stuff broke when the trailing slash was missing.

**Comment by PVince81** [comment]
Code looks good so far with minor issues.

Next up: smash this.

**Comment by PVince81** [comment]
To test:
- [ ] TEST: regular OC usage
- [ ] TEST: share, check recipient, check DB has the mount
- [ ] TEST: unshare, check recipient, check DB has removed the mount
- [ ] TEST: external storage mount/unmount
- [ ] TEST: manually clear oc_mount to simulate "old" cache, see how UI behaves

**Comment by PVince81** [comment]
For file shares the mount point appears with a trailing slash, @icewind1991 can you confirm that this is expected and harmless ?

**Comment by PVince81** [comment]
Also one strange thing is that when I setup an ext storage mount or just accepted an incoming remote share, the database doesn't seem to be updated directly. But it could also be a small delay due to transactions because I had smashbox running in parallel and hammering the DB.

**Comment by PVince81** [comment]
- [ ] BUG: renaming mounts (see test cases above) does not update oc_mounts
- [ ] BUG: sometimes mounting a personal ext storage or incoming share doesn't seem to update the DB, even though I can access it in the web UI. Maybe it's because a reused an existing storage ? (re-mounted after deletion)

**Comment by PVince81** [comment]
I got the upload test fails again with smashbox because I ran out of space... I'll rerun it once the errors above are fixed.

File: lib/private/files/config/usermountcache.php (hunk 8 above)
Comment (icewind1991) [inline lib/private/files/config/usermountcache.php]: afaik that doesn't work inside the values array

File: lib/private/files/config/usermountcache.php (hunk 9 above)
Comment (icewind1991) [inline lib/private/files/config/usermountcache.php]: no, duplicated values are not a problem since it's the same data anyway

File: lib/public/files/config/imountprovidercollection.php (hunk 12 above)
Comment (icewind1991) [inline lib/public/files/config/imountprovidercollection.php]: used in phpdoc

File: tests/lib/files/config/usermountcache.php (hunk 16 above)
Comment (icewind1991) [inline tests/lib/files/config/usermountcache.php]: > Does creating a mount without trailing slash auto-adds the trailing slash

Yes

**Comment by PVince81** [comment]
I ran smashbox again during lunch excluding the test known to fail. All passed.

File: lib/private/files/filesystem.php (hunk 10 above)
Comment (icewind1991) [inline lib/private/files/filesystem.php]: Only other place it's used it mounting the root storage and we dont need to cache that one

**Comment by icewind1991** [comment]
> Also one strange thing is that when I setup an ext storage mount or just accepted an incoming remote share, the database doesn't seem to be updated directly

It get's updated the next time the fs is setup for the user

**Comment by icewind1991** [comment]
> TEST: delete user clears oc_mounts :no_entry_sign: FAIL

Fixed

**Comment by PVince81** [comment]
@icewind1991 did you fix all failed cases or should I wait before retesting ?

**Comment by DeepDiver1975** [comment]
@icewind1991 needs rebase

**Comment by DeepDiver1975** [comment]
rebased

**Comment by PVince81** [comment]
@icewind1991 see failing manual tests: https://github.com/owncloud/core/pull/20768#issuecomment-161265529

Renaming/moving mount points doesn't work and changing "applicable" doesn't update the oc_mounts table either.

**Comment by icewind1991** [comment]
Now handles moving mountpoints

Note that when changing mountpoints/applicables from settings the changes are only reflected in the mount cache after the user's filesystem is setup again

**Comment by PVince81** [comment]
- [ ] BUG: change ext storage applicable from "all" to "user1", then login as "user1" then "user2". The oc_mounts still contains the entries for all users, even "user2".

**Comment by PVince81** [comment]
Most cases are solved, only applicable has some trouble.

I still feel uneasy about the fact that the oc_mounts table only gets updated during the next setupFS.
The reason is that at some point we intend to have APIs for cron jobs to access this table to be able to "find user based on a file id". To be able to do so without setupFS, the table needs to be up to date at this stage. @icewind1991 @nickvergessen what do you think ?

**Comment by nickvergessen** [comment]
Well if the method does not return the actual user, my app will simply not work at all. For anyone not using the app, that might be okay. For others there should be something to force it (when possible).

**Comment by icewind1991** [comment]
We could add a cronjob to keep the table up to date

**Comment by PVince81** [comment]
> We could add a cronjob to keep the table up to date

But that means once again iterating over all users.
Or have a background job scheduled whenever such operation occur ?

Ideal would be an immediate update of the table, when possible. I know there are some cases where it might not be possible when affecting several users (the "applicable" case)

**Comment by icewind1991** [comment]
cache is now cleared for applicables that are removed (only removed, added it more complex)

**Comment by icewind1991** [comment]
@PVince81 @Xenopathic @DeepDiver1975 please review

**Comment by PVince81** [comment]
@icewind1991 I guess you didn't add the "update cache table directly after an update" yet ?
Let's do this separately then, as part of adding the API to resolve a user from a file id.

**Comment by icewind1991** [comment]
> Let's do this separately then

Would prefer that yes

**Comment by PVince81** [comment]
Changing applicable removes the entry now :+1:

**Comment by icewind1991** [comment]
oracle has actually proper failures for once, looking into them

**Comment by PVince81** [comment]
Second reviewer ? @rullzer @nickvergessen @DeepDiver1975 @Xenopathic

**Comment by PVince81** [comment]
@nickvergessen mind doing the second review ?

**Comment by PVince81** [comment]
@Xenopathic second review ?

File: lib/private/files/config/cachedmountinfo.php [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+	 */
+	public function getRootId() {
+		return $this->rootId;
+	}
+
+	/**
+	 * @return Node the root node of the mount
+	 */
+	public function getMountPointNode() {
+		// TODO injection etc
```
Comment (DeepDiver1975) [inline lib/private/files/config/cachedmountinfo.php]: @icewind1991 follow up pr on this one - I assume

File: lib/private/files/config/usermountcache.php [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+				'mount_point' => ':mount'
+			]);
+		$query->setParameters([
+			':storage' => $mount->getStorageId(),
+			':root' => $mount->getRootId(),
+			':user' => $mount->getUser()->getUID(),
+			':mount' => $mount->getMountPoint()
+		]);
+		try {
+			$query->execute();
```
Comment (DeepDiver1975) [inline lib/private/files/config/usermountcache.php]: what about using insert if not exists?

File: lib/private/files/config/usermountcache.php (hunk 8 above)
Comment (nickvergessen) [inline lib/private/files/config/usermountcache.php]: It should, and if not try createParameter()

File: lib/private/files/config/usermountcache.php (hunk 18 above)
Comment (icewind1991) [inline lib/private/files/config/usermountcache.php]: done

File: lib/private/files/config/cachedmountinfo.php (hunk 17 above)
Comment (icewind1991) [inline lib/private/files/config/cachedmountinfo.php]: ye

File: lib/public/files/config/iusermountcache.php [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+ *
+ */
+
+namespace OCP\Files\Config;
+
+use OCP\Files\Mount\IMountPoint;
+use OCP\IUser;
+
+/**
+ * Cache mounts points per user in the cache so we can easilly look them up
```
Comment (MorrisJobke) [inline lib/public/files/config/iusermountcache.php]: easily

Comment (icewind1991) [inline lib/public/files/config/iusermountcache.php]: fixed

**Comment by PVince81** [comment]
@DeepDiver1975 can we merge this then ?

**Comment by DeepDiver1975** [comment]
:+1:
